lib/backend/watchersyncer/watchersyncer.go

```diff
@@ -0,0 +1,200 @@
+}
+
+// run implements the main syncer loop that loops forever receiving watch events and translating
+// to syncer updates.
+func (rs *watcherSyncer) run() {
+	log.Info("Running - sending status update")
+	rs.sendStatusUpdate(api.WaitForDatastore)
+	log.Info("Starting watcher caches")
+	for _, wc := range rs.watcherCaches {
+		go wc.run()
+	}
+
+	log.Info("Starting main watch processing loop")
+	var updates []api.Update
+	for {
+		// Block until there is data.
+		result := <-rs.results
+
+		// Process the data - this will append the data in subsequent calls, and action
+		// it if we hit a non-update event.
+		updates := rs.processResult(updates, result)
+
+		// Append results into the one update until we either flush the channel or we
+		// hit our fixed limit per update.
+	consolidatationloop:
+		for ii := 0; ii < maxUpdatesToConsolidate; ii++ {
+			select {
+			case next := <-rs.results:
+				updates = rs.processResult(updates, next)
+			default:
+				break consolidatationloop
+			}
+		}
+
+		// Perform final processing (pass in a nil result) before we loop and hit the blocking
+		// call again.
+		updates = rs.processResult(updates, nil)
+	}
+
+}
+
+// Process a result from the result channel.  We don't immediately action updates, but
+// instead start grouping them together so that we can send a larger single update to
+// Felix.
+func (rs *watcherSyncer) processResult(updates []api.Update, result interface{}) []api.Update {
+
+	// Final call of the interation, send any updates if we've collated some.
+	if result == nil {
+		if len(updates) > 0 {
+			rs.callbacks.OnUpdates(updates)
+		}
+		return nil
+	}
+
+	// Switch on the result type.
+	switch r := result.(type) {
+	case []api.Update:
+		// This is an update.  If we don't have previous updates then also check to see
+		// if we need to shift the status into Resync.
+		// We append these updates to the previous if there were any.
+		if len(updates) == 0 {
+			if rs.status == api.WaitForDatastore {
+				rs.sendStatusUpdate(api.ResyncInProgress)
+			}
+			updates = r
+		} else {
+			updates = append(updates, r...)
+		}
+
+	case error:
+		// Received an error.  Firstly, send any updates that we have grouped.
+		if len(updates) > 0 {
+			rs.callbacks.OnUpdates(updates)
+			updates = nil
+		}
+
+		// If this is a parsing error, and if the callbacks support
+		// it, then send the error update.
+		log.WithError(r).Info("Error received in syncer")
+		if ec, ok := rs.callbacks.(api.SyncerParseFailCallbacks); ok {
+			if pe, ok := r.(cerrors.ErrorParsingDatastoreEntry); ok {
+				ec.ParseFailed(pe.RawKey, pe.RawValue)
+			}
+		}
+
+	case api.SyncStatus:
+		// Received a synced event.  If we are still waiting for datastore, send a
+		// ResyncInProgress since at least one watcher has connected.
+		log.Info("Received sync event from watcher")
+		if rs.status == api.WaitForDatastore {
+			rs.sendStatusUpdate(api.ResyncInProgress)
+		}
+
+		// Increment the count of synced events.
+		rs.numSynced++
```

Comment:
The individual watcher caches send an `api.InSync` every time they do a resync. This means that if one of them runs in "polling mode", or even if running in regular watcher mode, we might get more than one sync status update from a given watcherCache before all of the others have finished syncing.

The logic as it is now will handle this poorly - I think we probably need a map which tracks whether or not each watcherCache is actually in sync, rather than assuming that X messages means X watcherCaches are in sync.

Reply:
might need to add an interface on top of the watcher cache with an `InSync` method so that we can tell when each one is in sync.

Reply:
Not quite.  Each individual sub-synced will only send an InSync message once.  So when stuck in a polling loop - you will only get one sync message from that sub syncer.

I'm pretty sure I test that case too.

Reply:
Ah yeah, you're right - I missed that.

Looks like we never set that back to false too, so there's no way we'll get multiple of those.